gui/imagealbum-dialog.cpp

```diff
@@ -0,0 +1,463 @@
+	ButtonWidget *_prevButton;
+	ButtonWidget *_nextButton;
+	ButtonWidget *_saveButton;
+
+	StaticTextWidget *_imageNumberLabel;
+
+	ContainerWidget *_imageContainer;
+	GraphicsWidget *_imageGraphic;
+
+	bool _canAlwaysSaveImage;
+
+	enum {
+		kPrevCmd = 'PREV',
+		kNextCmd = 'NEXT',
+		kSaveCmd = 'SAVE',
+	};
+};
+
+
+ImageAlbumDialog::ImageAlbumDialog(const Common::U32String &title, ImageAlbumImageSupplier *imageSupplier, uint initialSlot)
+	: Dialog("ImageAlbum"), _imageSupplier(imageSupplier), _currentSlot(initialSlot), _numSlots(0), _imageGraphic(nullptr), _canAlwaysSaveImage(false) {
+
+	_backgroundType = ThemeEngine::kDialogBackgroundSpecial;
+
+	_numSlots = imageSupplier->getNumSlots();
+
+	assert(_numSlots > 0);
+
+	new StaticTextWidget(this, "ImageAlbum.Title", title);
+
+	if (initialSlot >= _numSlots)
+		initialSlot = _numSlots - 1;
+
+	// Buttons
+	if (_numSlots > 1) {
+		_prevButton = new ButtonWidget(this, "ImageAlbum.Prev", _("Prev"), Common::U32String(), kPrevCmd);
+		_nextButton = new ButtonWidget(this, "ImageAlbum.Next", _("Next"), Common::U32String(), kNextCmd);
+		_imageNumberLabel = new StaticTextWidget(this, "ImageAlbum.ImageNumber", Common::U32String());
+	} else {
+		_prevButton = nullptr;
+		_nextButton = nullptr;
+		_imageNumberLabel = nullptr;
+	}
+
+	_saveButton = nullptr;
+	if (g_system->hasFeature(OSystem::kFeatureSystemSaveFileDialog)) {
+
+		Common::DialogManager *dialogManager = g_system->getDialogManager();
+
+		bool canSaveAnyFormat = false;
+		for (uint fmtID = Common::FileFormat::kFirstImageFormat; fmtID <= Common::FileFormat::kLastImageFormat; fmtID++) {
+			Common::FileFormat::Format format = static_cast<Common::FileFormat::Format>(fmtID);
+
+			if (dialogManager->getSaveFormatSupportLevel(format) > Common::FileFormat::kFormatSupportLevelNone) {
+				canSaveAnyFormat = true;
+				if (Common::FileFormat::getImageSaveFunction(format) != nullptr) {
+					_canAlwaysSaveImage = true;
+					break;
+				}
+			}
+		}
+
+		if (canSaveAnyFormat) {
+			_saveButton = new ButtonWidget(this, "ImageAlbum.Save", _("Save Image..."), Common::U32String(), kSaveCmd);
+			_saveButton->setEnabled(!_canAlwaysSaveImage);
+		}
+	}
+
+	new ButtonWidget(this, "ImageAlbum.Close", _("Close"), Common::U32String(), kCloseCmd);
+
+	_imageContainer = new ContainerWidget(this, "ImageAlbum.ImageContainer");
+
+	_imageGraphic = nullptr;
+}
+
+ImageAlbumDialog::~ImageAlbumDialog() {
+}
+
+void ImageAlbumDialog::open() {
+	Dialog::open();
+
+	changeToSlot(_currentSlot);
+}
+
+void ImageAlbumDialog::changeToSlot(uint slot) {
+	ContainerWidget *container = _imageContainer;
+	container = _imageContainer;
+
+	bool canSaveImage = _canAlwaysSaveImage;
+
+	if (_imageGraphic) {
+		_imageContainer->removeWidget(_imageGraphic);
+		delete _imageGraphic;
+		_imageGraphic = nullptr;
+	}
+
+	Common::Rect graphicRect = Common::Rect(0, 0, _imageContainer->getWidth(), _imageContainer->getHeight());
+
+	int inset = static_cast<int>(16.0f * g_gui.getScaleFactor());
```

Comment:
What is this magic number 16? It is best to put it to the theme file.

Reply:
Moved to theme file